larq/optimizers_v2.py

```diff
@@ -5,88 +5,75 @@
-            )
-
-        self.optimizer = optimizer
-        self.mask_fn = mask_fn
-
-    @staticmethod
-    def is_binary(var):
-        return "/kernel" in var.name and "quant_" in var.name
-
-
 @utils.register_keras_custom_object
-class OptimizerGroup(tf.keras.optimizers.Optimizer):
-    """A group of `MaskedOptimizer` that collectively train a model's variables.
+class CaseOptimizer(tf.keras.optimizers.Optimizer):
+    """A case operation for optimizers that each train a subset of a model's variables.
 
-    The set of `mask_fn` passed through the list of `MaskedOptimizer` should partition
-    the model's variables, so that exactly one optimizer is responsible for training 
-    each variable. The last `MaskedOptimizer` in the list may have its `mask_fn` set to
-    `None`; this is the default optimizer that will train any variables not claimed by
-    any previous optimizer.
+    An optimizer is used to train a variable iff its accompanying predicate evaluates to
+    `True`.
+
+    For each variable, at most one optimizer's predicate may evaluate to `True`. If no
+    optimizer's predicate evaluates to `True` for a variable, it is trained with the
+    `default` optimizer.
 
     # Arguments
-    masked_optimizers: a list of `MaskedOptimizer`
+    pred_opt_pairs: a list of `(tf.keras.optimizers.Optimzer, pred)` pairs, where `pred` 
+        takes one `tf.Variable` as argument and returns `True` if the optimizer should
+        train that variable, e.g. `pred(var) === True`.
+    default: a `tf.keras.optimizers.Optimizer` to be applied to any variable not claimed
+        by any other optimizer.
     """
 
-    def __init__(self, masked_optimizers, name="Group"):
+    def __init__(self, pred_opt_pairs, default, name="optim_case"):
         super().__init__(name=name)
 
-        for i, masked_opt in enumerate(masked_optimizers):
-            if not isinstance(masked_opt, MaskedOptimizer):
+        for i, (pred, opt) in enumerate(pred_opt_pairs):
+            if not isinstance(opt, tf.keras.optimizers.Optimizer):
                 raise TypeError(
-                    f"Expected `MaskedOptimizer` at index {i} but got `{type(masked_opt)}`."
+                    f"Expected `tf.keras.optimizers.Optimizer` at `pred_opt_pairs[{i}][1]` but got `{type(opt)}`."
                 )
+        self.pred_opt_pairs = pred_opt_pairs
 
-        self.masked_optimizers = masked_optimizers
+        if (
+            not isinstance(default, tf.keras.optimizers.Optimizer)
+            and default is not None
+        ):
+            raise TypeError(
+                f"Expected `tf.keras.optimizers.Optimizer` for `default` but got `{type(fallback_optimizer)}`."
+            )
+        self.default = default
 
     def __getattr__(self, name):
-        if name == "lr":  # TODO: Which optimizer's LR should handle this?
+        if name == "lr":  # TODO: Return list of learning rates
```

Comment:
Would raise an error, we can make this nice later if there is demand for it

Reply:
Implemented throwing a `ValueError` for this in 151af08.